src/wp-includes/rest-api/endpoints/class-wp-rest-abilities-list-controller.php

```diff
@@ -0,0 +1,335 @@
+	 * @return WP_REST_Response Response object on success.
+	 */
+	public function get_items( $request ) {
+		$abilities = array_filter(
+			wp_get_abilities(),
+			static function ( $ability ) {
+				return $ability->get_meta_item( 'show_in_rest' );
+			}
+		);
+
+		// Filter by ability category if specified.
+		$category = $request->get_param( 'category' );
+		if ( ! empty( $category ) ) {
+			$abilities = array_filter(
+				$abilities,
+				static function ( $ability ) use ( $category ) {
+					return $ability->get_category() === $category;
+				}
+			);
+			// Reset array keys after filtering.
+			$abilities = array_values( $abilities );
+		}
+
+		// Handle pagination with explicit defaults.
+		$params   = $request->get_params();
+		$page     = $params['page'] ?? 1;
+		$per_page = $params['per_page'] ?? self::DEFAULT_PER_PAGE;
+		$offset   = ( $page - 1 ) * $per_page;
+
+		$total_abilities = count( $abilities );
+		$max_pages       = ceil( $total_abilities / $per_page );
+
+		if ( $request->get_method() === 'HEAD' ) {
+			$response = new WP_REST_Response( array() );
+		} else {
+			$abilities = array_slice( $abilities, $offset, $per_page );
+
+			$data = array();
+			foreach ( $abilities as $ability ) {
+				$item   = $this->prepare_item_for_response( $ability, $request );
+				$data[] = $this->prepare_response_for_collection( $item );
+			}
+
+			$response = rest_ensure_response( $data );
+		}
+
+		$response->header( 'X-WP-Total', (string) $total_abilities );
+		$response->header( 'X-WP-TotalPages', (string) $max_pages );
+
+		$query_params = $request->get_query_params();
+		$base         = add_query_arg( urlencode_deep( $query_params ), rest_url( sprintf( '%s/%s', $this->namespace, $this->rest_base ) ) );
+
+		if ( $page > 1 ) {
+			$prev_page = $page - 1;
+			$prev_link = add_query_arg( 'page', $prev_page, $base );
+			$response->link_header( 'prev', $prev_link );
+		}
+
+		if ( $page < $max_pages ) {
+			$next_page = $page + 1;
+			$next_link = add_query_arg( 'page', $next_page, $base );
+			$response->link_header( 'next', $next_link );
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Retrieves a specific ability.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param WP_REST_Request<array<string, mixed>> $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_item( $request ) {
+		$ability = wp_get_ability( $request->get_param( 'name' ) );
+		if ( ! $ability || ! $ability->get_meta_item( 'show_in_rest' ) ) {
+			return new WP_Error(
+				'rest_ability_not_found',
+				__( 'Ability not found.' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		$data = $this->prepare_item_for_response( $ability, $request );
+		return rest_ensure_response( $data );
+	}
+
+	/**
+	 * Checks if a given request has access to read abilities.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param WP_REST_Request<array<string, mixed>> $request Full details about the request.
```

Comment:
```suggestion
	 * @param WP_REST_Request $request Full details about the request.
```

The request object doesn't have generic typing.